### ISSUE TITLE
Please create a new release to help packaging in distros

### ISSUE THREAD

**Issue by dod38fr** [libtom/libtommath#35]
Hello

I'm trying to create a Debian package for Perl6. This packaging is currently blocked by a [test failure](https://github.com/perl6/nqp/issues/234). The root cause of this failure is fixed in libtommath by commit 6907f6ca4582064f28e3a58374083ab2a4d85418 (merged).

To provide a proper Perl6 package, this library must be provided by [libtommath Debian package](https://tracker.debian.org/pkg/libtommath) which is at version 0.42.0 (4 years old). 

I'd like to create a new version of libtommath package so that the above fix is available for Debian users (and Perl6 package).

To make this work much easier, could you create a new release of libtommath ?

All the best

**Comment by baby-gnu** [comment]
Hello,

Any news on the v0.43 release ?

Regards.

**Comment by tomstdenis** [comment]
Ugh, the "maintainers" are not keen on a release.  I am the Tom of LibTom ... fuck it.  Barring anything huge let's aim for a November 13th release of v0.43.  The code is fairly stable.

**Comment by dod38fr** [comment]
Good news. Thanks !

**Comment by sjaeckel** [comment]
@tomstdenis now that you started it and brought up a date: what is still missing that we missed the target?
I've marked all relevant issues & PR's to https://github.com/libtom/libtommath/milestones/v0.43.0
Can you please add issues for stuff that should be done I didn't think of yet?

**Comment by tomstdenis** [comment]
Honestly, the only thing that sticks out to me is that some of the CLANG
-m32 tests fail on my F23 box.  I'd like to drop clang from the test suite
(or make it optional).

Beyond that I'd be ok with making tarballs anyday now.

**Comment by sjaeckel** [comment]
For me also gcc -m32 tests fail IIRC

**Comment by tomstdenis** [comment]
running right now -m32 passes with gcc only clang fails.   Though my setup
is a bit weird since I have a nightly build of clang/etc not the F23
install.

Frankly right now I'm ok with leaving it at gcc when testing and we can
circle back to clang/etc later.

**Comment by tomstdenis** [comment]
-mx32 fails for me but I probably don't have X32 libs installed (nor do I
care to right now).

Focus if anything should be on getting LTM 0.43 out and then updates to
TFM/LTC.

**Comment by sjaeckel** [comment]
```
$ COMPILERS=gcc-5 ./testme.sh                                                                                                  
uname=Linux hp-laptop 3.13.0-44-generic #73-Ubuntu SMP Tue Dec 16 00:22:43 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
gcc-5=5.2.1
Run test gcc-5 
Run test gcc-5 -DMP_8BIT
Run test gcc-5 -DMP_16BIT
Run test gcc-5 -DMP_32BIT
Run test gcc-5 -m32 
Run test gcc-5 -m32 -DMP_8BIT
error 124 while running tests
assuming timeout while running test - continue
Run test gcc-5 -m32 -DMP_16BIT
error 124 while running tests
assuming timeout while running test - continue
Run test gcc-5 -m32 -DMP_32BIT
Run test gcc-5 -mx32 
Run test gcc-5 -mx32 -DMP_8BIT
Run test gcc-5 -mx32 -DMP_16BIT
Run test gcc-5 -mx32 -DMP_32BIT
clang=4.2.1
Run test clang 
Run test clang -DMP_8BIT
Run test clang -DMP_16BIT
Run test clang -DMP_32BIT
Run test clang -m32 
Run test clang -m32 -DMP_8BIT
Run test clang -m32 -DMP_16BIT
error 124 while running tests
assuming timeout while running test - continue
Run test clang -m32 -DMP_32BIT
gcc=4.9.3
Run test gcc 
Run test gcc -DMP_8BIT
Run test gcc -DMP_16BIT
error 124 while running tests
assuming timeout while running test - continue
Run test gcc -DMP_32BIT
Run test gcc -m32 
Run test gcc -m32 -DMP_8BIT
error 124 while running tests
assuming timeout while running test - continue
Run test gcc -m32 -DMP_16BIT
error 124 while running tests
assuming timeout while running test - continue
Run test gcc -m32 -DMP_32BIT
Run test gcc -mx32 
Run test gcc -mx32 -DMP_8BIT
Run test gcc -mx32 -DMP_16BIT
Run test gcc -mx32 -DMP_32BIT
6 tests timed out
$ git describe --tags --always 
v0.42.0-149-g10ba45b
```

**Comment by sjaeckel** [comment]
I just did some tests and it looks like  `Testing (not safe-prime):        46 bits` when compiled with `MP_16BIT` is always the reason for the failure on 32bit architectures.
I also tested on ARM 32bit with gcc 4.9.2 and 5.1.1 and the same issue occurs there.
It also doesn't look like the entropy from rand() is responsible as testing with `LTM_DEMO_REAL_RAND` defined also fails.

**Comment by dod38fr** [comment]
Hello

The tests of the new version of Perl6 [fail again](https://github.com/perl6/nqp/issues/269) due to another issue with libtommath. This issue is fixed in develop branch with 265651b2c7e7800c8e8d257057d4494d78bded95.

@tomstdenis : what's the status of 0.43.0 release ?  

If this release is not out soon, I'll have to create another patch for Debian's version of libtommath.

All the best

**Comment by TazeTSchnitzel** [comment]
I hope we can get 0.43.0 out at some point. There's lots of useful stuff that's been added since the previous release; if I ever use libtommath again, I'll want to use it.

**Comment by sjaeckel** [comment]
As soon as #43 is merged we can do a RC, so @dfateyev and probably also @stapelberg can start a testrun for Fedora resp. Debian.
After the persons who ACK'ed testing the RC gave the okay it can be released.

**Comment by sjaeckel** [comment]
Oh I just realized that @stapelberg is only maintainer of ltc, well then I'll have to write an email about the RC to Joe Nahmias <jello@debian.org>

**Comment by dod38fr** [comment]
@sjaeckel : please keep me in the loop for Debian RC test. I'm also de Debian Developer (dod@debian.org) and I can help testing on Debian.

All the best

**Comment by tomstdenis** [comment]
Hey guys, jumping in here again.  Just to let you know I'm following the threads but my free time is basically 0 nowadays.  We really should get an LTM release out the door though.  I'll have some free time this weekend.  What are the major blockers people can't kick to 0.44?

**Comment by sjaeckel** [comment]
Good to hear you're still there :-)
I think after merging #43 and #47 we should be fine to go.

**Comment by dfateyev** [comment]
From my side there are no objections.
There are still some issues with building PDF docs, but they are pure Fedora and RHEL-specific. I think we'll skip docs from build if don't find a solution (not critical, in other words).

I'm for the release as soon as possible. But I'd like to test the final RC version before release.

**Comment by tomstdenis** [comment]
I can build the docs in Fedora 23.

I can run the tests and the docs build and the "zipup" target is modern and functional.  I'll be pushing a v0.43-rc0 tag in a minute.  If there are no complaints I'll push v0.43 on Feb 5th 2016.

**Comment by sjaeckel** [comment]
can we please stick with the 0.43.0 pattern...

**Comment by tomstdenis** [comment]
Sure, I'll repush the tag.  Note that I don't intend to release point
releases ... next will be v0.44.0

**Comment by sjaeckel** [comment]
TBH I'd prefer to have a proper version scheme especially in regards to .so versions.
Which also means that ltm 0.43.0 .so version will require a bump of the major version as the ABI of the library will change!
...and I just realized that this means that 8fde8fa41b89d24d45e9487ad49ba5d3c2fa5a76 was pointless, or we'd have to go for v1.0.0

**Comment by tomstdenis** [comment]
To be honest I so rarely used shared objects (in my code...) that that hasn't occurred to me.  I've spent the last 10+ years of my life buried in microcontrollers that shared objects don't exist :-)

Let's call the Feb 5th release v0.43.0 and then we can sort out version numbers later.  The ABI shouldn't have changed between 0.42 and 0.43 unless mp_int has changed?  The function prototypes should be stable.

**Comment by sjaeckel** [comment]
sadly that's an ABI change as well :-(

```
-char *mp_error_to_string(int code);
+const char *mp_error_to_string(int code);
```

**Comment by tomstdenis** [comment]
Well given the last release was a decade or so ago ... I think an ABI break is fine.  Just don't let it happen again :-)

I'm also ok if you want to call v0.43.0 v1.0.0 instead.  Just let me know before Feb 5th as we'll have to edit .tex and makefiles

**Comment by dfateyev** [comment]
> I can build the docs in Fedora 23.

It's broken in Fedora devel (rawhide) and partially in RHEL 7. But as said, there is not a major issue and can be sorted / worked around.

As for the .so versioning, I think we'll find a solution even prior the current release (will look at it tomorrow).

**Comment by tomstdenis** [comment]
Well any reason the RPM can't be built from an official tarball?  It would contain pdf manuals...

As for versioning.  I'd like to just keep bumping e.g. v0.43[.0].  If that's going to create some massive SO problem v1.0.0 is fine too.  Likely TFM will have a similar jump 0.13.0 => 1.0.0

**Comment by sjaeckel** [comment]
Regarding the versioning, I think we should just bump both, the regular and the .so version number, so everyone sees that there were major changes.

I also don't think that there will be major ABI/API changes (not additions) for LTM.

For TFM I also guess that the next release would be 1.0.0

**Comment by TazeTSchnitzel** [comment]
Given the gap since the last release, jumping to `v1.0.0` seems fitting.

**Comment by dfateyev** [comment]
> Well any reason the RPM can't be built from an official tarball? It would contain pdf manuals...

Yeah, I think we'll switch to this approach. The current RPM spec builds PDF every time while the package is being built.
Then manuals should always be present in tarball (now some are missing).

**Comment by sjaeckel** [comment]
Which PDF's are missing from the latest release tarball?
Most PDF's aren't tracked anymore in the repository, that's correct, but v0.42.0 should contain all of the relevant files.

**Comment by dfateyev** [comment]
Right, 0.42.0 does have all, and in the current version at least "bn.pdf" and "poster.pdf" are missing.

**Comment by tomstdenis** [comment]
I just made a 0.43 tarball and it had all three pdfs.

**Comment by sjaeckel** [comment]
@dfateyev because you're using the tarballs that are created by Github from the repository!?!

**Comment by tomstdenis** [comment]
"make zipup" makes a tar.xz file that contains the PDFs.  Distros should make installable packages from that tarball alone.

**Comment by dfateyev** [comment]
> Distros should make installable packages from that tarball alone.

I meant the best approach that Github tarball would contain all PDFs out of box.
Along already presented "tommath.pdf" which perhaps needs to be updated.

**Comment by tomstdenis** [comment]
I'm not super in favour of committing pdfs to the repo because it bulks up the repo size somewhat for not much added value.  Having the PDFs in the release tarballs allows us to preserve the renders.

**Comment by TazeTSchnitzel** [comment]
Not only do they bulk up the size, but it means we have duplicate information that can go out-of-sync. Can we drop the existing PDF in the repo?

**Comment by tomstdenis** [comment]
Removing the PDF today won't shrink the repo size unless we do a filter branch which I don't want to.  I'd be ok with doing a "git rm" if you want but only to not leave out of date pdfs in the work tree.

**Comment by TazeTSchnitzel** [comment]
Yes, `git rm` is what I mean. Obviously the current and older versions will remain in git history, but we shouldn't repeat our past mistakes.

**Comment by dfateyev** [comment]
>  Having the PDFs in the release tarballs allows us to preserve the renders.

If the release tarball contain them, sure I'm totally fine with it. A bit confused with currently tagged RCs https://github.com/libtom/libtommath/releases which don't contain them.

**Comment by TazeTSchnitzel** [comment]
At the very least `ltm-0.42.0.tar.bz2` contains the PDFs. Are you downloading the provided tarballs/zips, or are you clicking GitHub's "Source code" option?

**Comment by tomstdenis** [comment]
In an ideal world you'd use the git tree based on the release tag and just run make "docs poster manual" to make the PDFs.

In a less ideal world you'd use a tarball we provide based on the release tag which already includes them and the source tree.

**Comment by dfateyev** [comment]
@TazeTSchnitzel v0.42.0 is fine. I'm looking at tagged RC (at least they're tagged that way): 
![pic10](https://cloud.githubusercontent.com/assets/727201/12704612/0bcaa064-c889-11e5-8078-23f092214364.png)

**Comment by TazeTSchnitzel** [comment]
Oh, I see. Yes, you're right. It looks like there's no actual GitHub "release" there, GitHub has just taken the tags and packaged the source code.

**Comment by tomstdenis** [comment]
The PDFs won't be committed to the tree though it's a waste of space.  The release tarballs will have them.

Remember the packaging for these predate git and github by quite a bit.

**Comment by dfateyev** [comment]
@sjaeckel  considering 1.x.x versioning schema the changes in 8fde8fa definitely looks useless. Please simply revert this commit.
As for .so versioning, due to ABI change in the upcoming 1.0.0 release we should increase "current" tag which is currently 0. As for "revision" it will automatically be "0". 
`VERSION=1.0.0` `VERSION_SO=1:0` if it's fine to everybody.

**Comment by tomstdenis** [comment]
I have no qualms with that change.  I'm not a big fan of the 3 tiers though.  I can barely support a release let alone multiple.  There will never be a 1.0.x for x>0 since I don't plan on supporting say 1.0.1 and 1.1 and ... If we fix bugs in 1.0 the next release will be 1.1.  It's enough to indicate a higher version than the previous.  

Users noticing a bug or needing a feature simply have to upgrade or produce their own package.

**Comment by sjaeckel** [comment]
FWIW, I simply don't see the need to drop the 0 away that costs us nothing but lets us in case there's only a bugfix and no feature addition to just replace the .0 by a .1 and push that release out.

1. I may remember you of v0.10a and v0.10b of tfm IIRC ... yes you also did bugfix releases...
2. I simply don't like it to go my own way if there are established solutions that everyone knows how they work

**Comment by tomstdenis** [comment]
To be fair I was a one-man show back in the early days and was really
winging it.

Since I don't have the time to support multiple parallel versions (1.x.y
implies a branch btw) I'd rather stick to major/minor and simply have
develop be the latest code and tags along the way reflecting a version.

**Comment by sjaeckel** [comment]
Hey that doesn't mean that we have to support different minor versions for ages, but just in case there were no feature additions since the last release we could just stick with the established `major.minor.patch+1` syntax and everyone knows "I can just use it as a drop-in and it won't break my dependencies".

**Comment by sjaeckel** [comment]
I just think either we do it right or there's no need to even bother.

Until you came back to join the not-anymore-one-man-show I didn't bother and simply did no releases.
Now you want to do it right by doing a release, and then we should also do it entirely and not only half.

**Comment by TazeTSchnitzel** [comment]
I agree that we should keep around the third number. We probably won't use it, but we never know, and it can't hurt.

**Comment by tomstdenis** [comment]
Ya but then you end up with branches that sit dormant for years and when
someone does "git branch -a" they'll see 200 branches ...

Generally speaking one of the virtues of the LT projects was API changes
were rare.  In fact most of the LTM API is still compatible with MPI from
the 90s.

If it makes a distro guy happy to revert to maj/min then that's enough for
me.

**Comment by tomstdenis** [comment]
@TazeTSchnitzel @dfateyev claims maj/min helps the packaging on their end and I don't ever see us branching to do patches.  So I'm inclined to help the user who is trying to distribute the project widely :-)

**Comment by TazeTSchnitzel** [comment]
Fair enough. Also, I suppose we could always do `v1.0b` or whatever if we needed to.

**Comment by tomstdenis** [comment]
It's all "made up" anyways.  Linux does x.y but branch maintainer append on their own .z, etc... OpenSSL does well... who the fuck knows x.y+z of some sort.

I just want to make it clear that I don't intend to support branches :-) If 1.0 has a bug 1.1 will have the fix.  And it's not decimal either ... after 1.9 would just be 1.10 and so on.  If we make any sort of major API changes we can go for 2.x and so on.

To be brutally honest I also don't see a 2.x of LTM.  The idea was to be stable and portable.

Now TFM ... I see a 2.x of that when/if we move to multiple precision.  That will break all sort of ABI compatibility.  Frankly, it'd probably be wise to make 1.x be the release that is mp and just doddle along on 0.x with the fixed precision.

**Comment by sjaeckel** [comment]
> @dfateyev claims maj/min helps the packaging on their end and I don't ever see us branching to do patches.

I can't find that claim, but I probably oversaw it.

> It's all "made up" anyways

But therefor there's a lot of software depending on stuff like [1] or [2]

> I just want to make it clear that I don't intend to support branches :-)

As long as you don't start a support branch I doubt someone will think there will be older releases that will be supported :-)

> And it's not decimal either ... after 1.9 would just be 1.10 and so on

anything else than 1.10 after 1.9 would be ridiculous :-D

>  I also don't see a 2.x of LTM

I also don't see it and I think the API is now fine

> Now TFM ... I see a 2.x of that when/if we move to multiple precision. That will break all sort of ABI compatibility. Frankly, it'd probably be wise to make 1.x be the release that is mp and just doddle along on 0.x with the fixed precision.

sounds good, and if you hear the plans of @sebastianas about the implementation we could even stay with the same implementation for fixed and multiple precision.


[1] https://www.gnu.org/software/libtool/manual/libtool.html#Updating-version-info
[2] http://semver.org/

**Comment by tomstdenis** [comment]
@dfateyev had asked to revert the patch that introduced x.y.z versioning. 

The typical way to do a patch release is you branch the maj/min tag and then commit changes there.  Then you tag that and produce a release.  It means I/someone has to apply the fix to both the patch branch and the develop branch.

w.r.t. TFM.  From my experience the upsides of mp far outweigh fp except in very tight contrived cases.  I'd be ok with entertaining discussion about supporting both but keep in mind the switch is compile time realistically not runtime.

**Comment by dfateyev** [comment]
> @dfateyev had asked to revert the patch that introduced x.y.z versioning. 
??

>>    @dfateyev claims maj/min helps the packaging on their end and I don't ever see us branching to do patches.

> I can't find that claim, but I probably oversaw it.

Possibly this "`VERSION=1.0.0 VERSION_SO=1:0` if it's fine to everybody" has been interpreted like this ;-) Actually, I'm not a strictly supporter of 1.x.x scheme, and the versioning doesn't bother me unless it breaks things. The schema "1.x" is fine to me too, since 1.0 will be newer than 0.42.0.

As for `VERSION_SO=1:0` I believe that everybody knows that for sure, but just a clarification for everybody involved: it has nothing to do with `VERSION`. We're bumping from the current "0:43" to "1:0" to reflect ABI changes and get "libtommath.so.0" -> "libtommath.so.1". If any major changes with ABI occur in the future, it will be "2:0" ("libtommath.so.2"), "3:0" etc., regardless the `VERSION`.

**Comment by dfateyev** [comment]
To make long story short:
-  To me, both LTM versioning version are fine (1.x.x and 1.x), seems both are fine and vitable;
- The "current" tag in `VERSION_SO` should be bumped from 0 to 1, in other words, `VERSION_SO` should be "1:0" now (and it already is the repo);
- Before the release please prepare RC which containts everything identical to the release, to check everything out.

**Comment by tomstdenis** [comment]
The only difference between the release tarball and the tag is the pdfs.

**Comment by sjaeckel** [comment]
@dfateyev I just added an rc2 tag including the archives, please feel free to test as if it were the final release and report any issues.

**Comment by tomstdenis** [comment]
@sjaeckel You added gpg signatures back ... whose key is doing the signing?  I don't have a gpg key (and if I did nobody has signed it ...).  What are your thoughts there?

**Comment by sjaeckel** [comment]
I can finish the release and sign the archives.

**Comment by tomstdenis** [comment]
Ya, but like, uh, why do I trust your key?  :-)

**Comment by sjaeckel** [comment]
Because there's an incredible number of 2 signatures on it? ;-)

**Comment by tomstdenis** [comment]
Sold!

**Comment by dfateyev** [comment]
@sjaeckel RC2 looks fine. Built and slightly tested (including `./test && ./ltmtest` where applicable) on x86, amd64, armv7, aarch64, ppc64 and ppc64le; no issues found.
Please don't forget about 3 PDFs to be included to the release tarball.

**Comment by karel-m** [comment]
IMHO libtommath_VS2005* + libtommath_VS2008* files should be removed from the release tarball as libtommath cannot be built with MS Visual Studio 2008 or earlier (due to stdint.h).

**Comment by tomstdenis** [comment]
We could always add a user supplied "stdint.h" later on.  (I've done this
in proprietary projects before).

I don't run Windows anywhere so I can't test them.

Tom

**Comment by tomstdenis** [comment]
Any complaints on rc2?

**Comment by tomstdenis** [comment]
No comments... I pushed v1.0 go ahead and make tarballs and call the press.

**Comment by sjaeckel** [comment]
@dfateyev @dod38fr ping

@tomstdenis tarballs uploaded, you wanna call the press? For me it's bedtime now :-)
OT: do you know any of the remaining administrators of the libtom ML? There should probably be some more admins...

**Comment by dfateyev** [comment]
Everything looks good to me. Thanks to everybody involved for the release.

**Comment by TazeTSchnitzel** [comment]
🎉🎉🎉

**Comment by dod38fr** [comment]
Thanks a bunch. I've uploaded [libtommath-1.0-2](https://tracker.debian.org/pkg/libtommath) to Debian/unstable.

All the best
